Move IPC from battle-engine.js to simulator.js

taylor1791 opened the pull request:
This is my attempt at resolving #2655. Special thanks to Zarel for guiding me in the right direction.

This moves all the `process` stuff out of `battle-engine.js` and into `simulator.js`. It is worth pointing out that this includes `Battle.prototype.send` because it depends on `process`. I decided to make it an argument to construct. If there is a different preference, let me know.

Zarel commented on simulator.js:
```diff
@@ -13,11 +13,16 @@
 
+global.Config = require('./config/config.js');
+global.Tools = require('./tools.js').includeMods();
+global.toId = Tools.getId;
```
I'm not sure we want to override the globals here. We should probably only do this after detecting a child process.

Zarel commented on battle-engine.js:
```diff
@@ -2045,6 +1955,8 @@ Battle = (() => {
 
 		this.startingSeed = this.generateSeed();
+
+		this.send = send;
```
In theory you just need to have it call `PM.send` instead of passing around a callback.

taylor1791 replied:
`PM` is an instance of `ProcessManager`. In `simulator.js` it is called `SimulatorProcess`. It is not clear to me how `battle-engine.js` could access `SimulatorProcess.send` without passing around a callback or making `battle-engine.js` dependent on `process-manager.js`.

Zarel replied:
Good point.

Zarel replied:
I would at least like `send` to be an optional argument, so you wouldn't need to mess with tests.

taylor1791 replied:
Oh, I guess I could `global.PM = SimulatorProcess;`. Is that what you meant?

taylor1791 replied:
Please ignore my previous comment. I didn't see your responses.

Zarel commented on battle-engine.js:
```diff
@@ -5011,14 +4923,6 @@ Battle = (() => {
-	Battle.prototype.send = function (type, data) {
-		if (Array.isArray(data)) data = data.join("\n");
-		process.send(this.id + "\n" + type + "\n" + data);
-	};
```
This should probably also stay, even if it's nulled out.

taylor1791 commented:
I updated my MR with the following:

 + added a default implementation of `Battle.send`.
 + readded `global.Config` to `process.js` for `Config.simulatorprocesses`
 + made `SimulatorProcess.spawn();` only spawn in a non-child process.

Zarel commented:
oh, also, please post here agreeing to license your contributions under the MIT license

taylor1791 commented:
I agree to license all my contributions to Zarel/Pokemon-Showdown under the MIT license.
